packages/contracts/contracts/Router.sol

```diff
@@ -0,0 +1,220 @@
+  modifier onlyViaFactory() {
+    require(msg.sender == routerFactory, "ONLY_VIA_FACTORY");
+    _;
+  }
+
+  function init(
+    address _transactionManager,
+    uint256 _chainId,
+    address _routerSigner,
+    address _recipient,
+    address _owner
+  ) external onlyViaFactory {
+    transactionManager = ITransactionManager(_transactionManager);
+    chainId = _chainId;
+    routerSigner = _routerSigner;
+    recipient = _recipient;
+    transferOwnership(_owner);
+  }
+
+  function setRecipient(address _recipient) external onlyOwner {
+    recipient = _recipient;
+  }
+
+  function setSigner(address _routerSigner) external onlyOwner {
+    routerSigner = _routerSigner;
+  }
+
+  function addRelayerFee(uint256 amount, address assetId) external payable {
+    // Sanity check: nonzero amounts
+    require(amount > 0, "#RC_ARF:002");
+
+    // Transfer funds to contract
+    // Validate correct amounts are transferred
+    if (LibAsset.isNativeAsset(assetId)) {
+      require(msg.value == amount, "#RC_ARF:005");
+    } else {
+      require(msg.value == 0, "#RC_ARF:006");
+      LibAsset.transferFromERC20(assetId, msg.sender, address(this), amount);
+    }
+
+    // Emit event
+    emit RelayerFeeAdded(assetId, amount, msg.sender);
+  }
+
+  function removeRelayerFee(uint256 amount, address assetId) external onlyOwner {
+    // Sanity check: nonzero amounts
+    require(amount > 0, "#RC_RRF:002");
+
+    // Transfer funds from contract
+    LibAsset.transferAsset(assetId, payable(recipient), amount);
+
+    // Emit event
+    emit RelayerFeeRemoved(assetId, amount, msg.sender);
+  }
+
+  function removeLiquidity(
+    uint256 amount,
+    address assetId,
+    bytes calldata signature
+  ) external {
+    if (msg.sender != routerSigner) {
+      SignedRemoveLiquidityData memory payload = SignedRemoveLiquidityData({
+        amount: amount,
+        assetId: assetId,
+        chainId: chainId
+      });
+
+      address recovered = recoverSignature(abi.encode(payload), signature);
+      require(recovered == routerSigner, "#RC_RL:040");
+    }
+
+    return transactionManager.removeLiquidity(amount, assetId, payable(recipient));
+  }
+
+  function prepare(
+    ITransactionManager.PrepareArgs calldata args,
+    address routerRelayerFeeAsset,
+    uint256 routerRelayerFee,
+    bytes calldata signature
+  ) external payable returns (ITransactionManager.TransactionData memory) {
+    if (msg.sender != routerSigner) {
+      SignedPrepareData memory payload = SignedPrepareData({
+        args: args,
+        routerRelayerFeeAsset: routerRelayerFeeAsset,
+        routerRelayerFee: routerRelayerFee,
+        chainId: chainId
+      });
+
+      address recovered = recoverSignature(abi.encode(payload), signature);
+      require(recovered == routerSigner, "#RC_P:040");
+
+      // Send the relayer the fee
+      if (routerRelayerFee > 0) {
+        LibAsset.transferAsset(routerRelayerFeeAsset, payable(msg.sender), routerRelayerFee);
+      }
```

Comment:
Also need to require that the amount stored in the contract is greater than the relayer fee

Reply:
Technically I don't think we need this. If the amount stored in the contract is < relayer fee then the relayer would never call this method - they would simulate the call, see that it would reject onchain, and then refuse to call. Same difference, as a require statement would have the same effect.

Reply:
sure it would fail if you tried to transfer more than what existed on the contract, but require statements return much better error messages

Reply:
I think that check would add more gas right? Isn't it better to just let it fail.

Reply:
@LayneHaber what would the error message look like w/o the require? Would it not be clear that it failed for that reason? I think that holds the answer

Reply:
Wouldn't it just be `ERC20: transfer amount exceeds balance` ? In which case I think it's descriptive enough